### PR TITLE
Forgot_password & Reset password features completed

### PR DESCRIPTION
**Pull request by ramagopal80**
Capture1
Capture2
Capture3
Capture4

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -15,4 +15,20 @@ spring.jpa.hibernate.ddl-auto = update
...
+# Email Properties
+# add these new properties
+spring.mail.transport.protocol=smtp
+spring.mail.from.email=ramagopal80@gmail.com
+spring.mail.port=587
 
+
+# modify these properties with your credentials
+spring.mail.username=ramagopal80@gmail.com
+spring.mail.password=Architha2010!!
```
Comment (prashanthich) [inline src/main/resources/application.properties]: your password compromised, reset password as earliest.

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    @Autowired
+    private UserRepository userRepo;
+
+    @ModelAttribute
+    @GetMapping("/forgot_password")
+    public String showForgotPasswordForm() {
+        return "forgot_password";
+    }
+
+    @PostMapping("/forgot_password")
```
Comment (prashanthich) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java]: use hyphen instead of underscore in route.

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+        helper.setSubject(subject);
+
+        helper.setText(content, true);
+
+        mailSender.send(message);
+    }
+
+
+    @GetMapping("/reset_password")
```
Comment (prashanthich) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java]: use hyphen instead of underscore in route.

**Review by prashanthich — COMMENTED** [review]
Rama, added my comments.

File: src/main/resources/application.properties (hunk 1 above)
Comment (ramagopal80) [reply src/main/resources/application.properties]: yeah, I will

File: build.gradle [hunk 4]
```diff
@@ -1,7 +1,21 @@
+buildscript {
+	ext {
+		springBootVersion = '1.5.2.RELEASE'
+	}
+	repositories {
+		mavenCentral()
```
Comment (korina0430) [inline build.gradle:6]: is this needed?

File: src/main/java/org/launhcode/healthynutrition/controllers/AuthenticationController.java [hunk 5]
```diff
@@ -39,6 +39,8 @@ public User getUserFromSession(HttpSession session) {
 
     private static void setUserInSession(HttpSession session, User user) {
 
+
+//        session.setAttribute(USER_SESSION_KEY, user.getEmail());
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/AuthenticationController.java:43]: remove commented code please

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    public String processForgotPassword(Model model, HttpServletRequest request) {
+        String email = request.getParameter("email");
+        String token = RandomString.make(30);
+
+        try {
+            customerService.updateResetPasswordToken(token, email);
+            String resetPasswordLink = Utility.getSiteURL(request) + "/reset_password?token=" + token;
+            sendEmail(email, resetPasswordLink);
+            model.addAttribute("message", "We have sent a reset password link to your email. Please check.");
+
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:52]: minor: remove extra spaces here

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+        } catch (UserNotFoundException ex) {
+            model.addAttribute("error", ex.getMessage());
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:54]: here you could add a specific error message like "user not found in the system"

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+        return "forgot_password";
+    }
+
+    public void sendEmail(String recipientEmail, String link)
+            throws MessagingException, UnsupportedEncodingException {
+        MimeMessage message = mailSender.createMimeMessage();
+        MimeMessageHelper helper = new MimeMessageHelper(message);
+
+        helper.setFrom("contact@shopme.com", "Shopme Support");
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:72]: shopme support does not seem related to this app :p

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        helper.setSubject(subject);
+
+        helper.setText(content, true);
+
+        mailSender.send(message);
+    }
+
+
+    @GetMapping("/reset_password")
+    public String showResetPasswordForm(@Param(value = "token") String token, @org.jetbrains.annotations.NotNull Model model) {
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:94]: This not null annotaiton doesn't look right..

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        User user = customerService.getByResetPasswordToken(token);
+        model.addAttribute("token", token);
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:96]: why does the token need to be part of the model?

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+    @PostMapping("/reset_password")
+    public String processResetPassword(HttpServletRequest request, Model model) {
+        String token = request.getParameter("token");
+        String password = request.getParameter("password");
+
+        User user = customerService.getByResetPasswordToken(token);
+        model.addAttribute("title", "Reset your password");
+
+        if (user == null) {
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:114]: you already do this check in the get mapping, do you also need to do it here?

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        User user = userRepo.findByEmail(email);
+        if (user != null) {
+            user.setResetPasswordToken(token);
+            userRepo.save(user);
+        } else {
+            errors.rejectValue(email, "Could not find any customer with the email ");
+        }
+    }
+
+    public User getByResetPasswordToken(String token) {
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:138]: i think this function would be more clear if it was named lookupUserByResetPasswordToken

File: src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
+public class CustomerService {
+
+    @Autowired
+    private UserRepository userRepo;
+
+
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java]: minor: remove the extra space here

File: src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import org.launhcode.healthynutrition.data.UserRepository;
+import org.launhcode.healthynutrition.models.User;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java]: what does @Transactional do?

File: src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        if (user != null) {
+            user.setResetPasswordToken(token);
+            userRepo.save(user);
+        } else {
+            throw new UserNotFoundException("Could not find any user with the email " + email);
+
+        }
+    }
+
+    public User getByResetPasswordToken(String token) {
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java]: this function also exists in forgotPasswordController. you should only need it in one place

File: src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        return userRepo.findByResetPasswordToken(token);
+    }
+
+    public void updatePassword(User user, String newPassword) {
+        BCryptPasswordEncoder passwordEncoder = new BCryptPasswordEncoder();
+        String encodedPassword = passwordEncoder.encode(newPassword);
+        user.setPassword(encodedPassword);
+
+        user.setResetPasswordToken(null);
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/models/service/CustomerService.java]: nice!

File: src/main/java/org/launhcode/healthynutrition/models/service/Utility.java [hunk 17]
```diff
@@ -0,0 +1,10 @@
+package org.launhcode.healthynutrition.models.service;
+
+import javax.servlet.http.HttpServletRequest;
+
+public class Utility {
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/models/service/Utility.java]: Creating a class for this function seems like overkill, just add the method to the class its used in

File: src/main/resources/application.properties [hunk 18]
```diff
@@ -15,4 +15,20 @@ spring.jpa.hibernate.ddl-auto = update
 # stripped before adding them to the entity manager)
 spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.MySQL8Dialect
 
+# Email Properties
+# add these new properties
```
Comment (korina0430) [inline src/main/resources/application.properties:19]: remove this comment pls

File: src/main/resources/templates/login.html [hunk 19]
```diff
@@ -1,11 +1,9 @@
 <!DOCTYPE html>
 <html lang="en" xmlns:th="http://www.thymeleaf.org/">
-<meta charset="UTF-8">
-<meta http-equiv="X-UA-Compatible" content="IE=edge">
 <meta name="viewport" content="width=device-width, initial-scale=1">
 <link href="https://fonts.googleapis.com/css?family=Roboto:400,700" rel="stylesheet">
 <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
-<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css">
```
Comment (korina0430) [inline src/main/resources/templates/login.html:8]: is bootstrap not used on this page? why did you remove it?

File: src/main/resources/templates/register.html [hunk 20]
```diff
@@ -1,11 +1,8 @@
 <!DOCTYPE html>
 <html lang="en" xmlns:th="http://www.thymeleaf.org/">
-<meta charset="UTF-8">
 <meta http-equiv="X-UA-Compatible" content="IE=edge">
-<meta name="viewport" content="width=device-width, initial-scale=1">
 <link href="https://fonts.googleapis.com/css?family=Roboto:400,700" rel="stylesheet">
 <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
-<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css">
```
Comment (korina0430) [inline src/main/resources/templates/register.html:8]: is bootstrap not used on this page? why did you remove it?

File: build.gradle (hunk 4 above)
Comment (ramagopal80) [reply build.gradle:6]: yes, without it, build.gradle gives an error as that's a separate one for the buildscript.

File: src/main/resources/application.properties (hunk 18 above)
Comment (ramagopal80) [reply src/main/resources/application.properties:19]: thought would be useful for my future reference

File: src/main/resources/templates/login.html (hunk 19 above)
Comment (ramagopal80) [reply src/main/resources/templates/login.html:8]: there are 2 bootstrap links, either one works fine

File: src/main/resources/templates/register.html (hunk 20 above)
Comment (ramagopal80) [reply src/main/resources/templates/register.html:8]: there are 2 bootstrap links, either one works fine

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java (hunk 8 above)
Comment (ramagopal80) [reply src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:72]: from this site only I'm getting me reset password link..
image

**Comment by revathivarini** [comment]
Looks like sending mail to reset password. Its good overall

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+//import static org.launhcode.healthynutrition.models.service.Utility.getSiteURL;
+
+@Controller
+public class ForgotPasswordController {
+    @Autowired
+    private JavaMailSender mailSender;
+
+//    @Autowired
+//    private CustomerService customerService;
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:31]: remove commented code

File: src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java (hunk 8 above)
Comment (korina0430) [reply src/main/java/org/launhcode/healthynutrition/controllers/ForgotPasswordController.java:72]: Yeah, I'm saying you should change the name to something related to this app instead of copying/pasting from the tutorial

File: src/main/resources/templates/login.html (hunk 19 above)
Comment (korina0430) [reply src/main/resources/templates/login.html:8]: oh didn't see that thanks!

**Review by korina0430 — APPROVED** [review]
